<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <title>Генераторы</title>
    <link href="../../style.css" rel="stylesheet">
</head>
<body>
    <article>
        <h2>Генераторы</h2>
        <p>Обычные функции возвращают только одно-единственное значение (или ничего).</p>
        <p>Генераторы могут порождать (yield) множество значений одно за другим, по мере необходимости.
        Генераторы отлично работают с перебираемыми объектами и позволяют легко создавать потоки данных.</p>
    </article>

    <article>
        <h2>Функция-генератор</h2>
        <p>Для объявления генератора используется специальная синтаксическая конструкция: <code>function*</code>,
        которая называется «функция-генератор».</p>
        <p>Выглядит она так:</p>
        <code>
            function* generateSequence() {<br>
            &nbsp;&nbsp;yield 1;<br>
            &nbsp;&nbsp;yield 2;<br>
            &nbsp;&nbsp;return 3;<br>
            }
        </code>
        <p>Функции-генераторы ведут себя не так, как обычные. Когда такая функция вызвана, она не выполняет свой код.
        Вместо этого она возвращает специальный объект, так называемый «генератор», для управления ее выполнением.</p>
        <p>Вот, посмотрите:</p>
        <code>
            function* generateSequence() {<br>
            &nbsp;&nbsp;yield 1;<br>
            &nbsp;&nbsp;yield 2;<br>
            &nbsp;&nbsp;return 3;<br>
            }<br>
            <br>
            // Функция-генератор создает объект генератор<br>
            let generator = generateSequence();<br>
            alert(generator); // [object Generator]
        </code>
        <p>Основным методом генератора является <code>next()</code>. При вызове он запускает выполнение кода до
        ближайшей инструкции <code>yield &ltзначение&gt</code> (значение может отсутствовать, в этом случае оно
        предполагается равным <code>undefined</code>). По достижении <code>yield</code> выполнение функции приостанавливается,
        а соответствующее значение - возвращается во внешний код:</p>
        <p>Результатом метода <code>next()</code> всегда является объект с двумя свойствами:</p>
        <ul>
            <li>
                <p><code>value</code>: значение из <code>yield</code>.</p>
            </li>
            <li>
                <p><code>done</code>: <code>true</code>, если выполнение функции завершено, иначе <code>false</code>.</p>
            </li>
        </ul>
        <p>Например, здесь мы создаем генератор и получаем первое из возвращаемых им значений:</p>
        <code>
            function* generateSequence() {<br>
            &nbsp;&nbsp;yield 1;<br>
            &nbsp;&nbsp;yield 2;<br>
            &nbsp;&nbsp;return 3;<br>
            }<br>
            <br>
            let generator = generateSequence();<br>
            <br>
            let one = generator.next();<br>
            <br>
            alert(JSON.stringify(one)); // {"value": 1, "done": false}
        </code>
        <p>На данный момент мы получили только первое значение, выполнение функции остановлено на второй строке.</p>
        <p>Повторный вызов <code>generator.next()</code> возобновит выполнение кода и вернет результат следующего
        <code>yield</code>:</p>
        <code>
            let two = generator.next();<br>
            <br>
            alert(JSON.stringify(two)); // {"value": 2, "done": false}
        </code>
        <p>И, наконец, последний вызов завершит выполнение функции и вернет результат <code>return</code>:</p>
        <code>
            let three = generator.next();<br>
            <br>
            alert(JSON.stringify(three)); // {"value": 3, "done": true}
        </code>
        <p>Сейчас генератор полностью выполнен. Мы можем увидеть это по свойству <code>done: true</code> и обработать
        <code>value: 3</code> как окончательный результат.</p>
        <p>Новые вызовы <code>generator.next()</code> больше не имеют смысла. Впрочем, если они и будут, то не вызовут
        ошибки, но будут возвращать один и тот же объект: <code>{done: true}</code>.</p>

        <article class="note">
            <h2><code>function* f(...)</code> или <code>function *f(...)</code>?</h2>
            <p>Нет разницы, оба синтаксиса корректны.</p>
            <p>Но обычно предпочтителен первый вариант, так как звездочка относится к типу объявляемой сущности
            (<code>function*</code> - «функция-генератор»), а не к ее названию, так что резонно расположить ее
            у слова <code>function</code>.</p>
        </article>
    </article>

    <article>
        <h2>Перебор генераторов</h2>
        <p>Как вы, наверное, уже догадались по наличию метода <code>next()</code>, генераторы являются перебираемыми
        объектами.</p>
        <p>Возвращаемые ими значения можно перебирать через <code>for...of</code>:</p>
        <code>
            function* generateSequence() {<br>
            &nbsp;&nbsp;yield 1;<br>
            &nbsp;&nbsp;yield 2;<br>
            &nbsp;&nbsp;return 3;<br>
            }<br>
            <br>
            let generator = generateSequence();<br>
            <br>
            for (let value of generator) {<br>
            &nbsp;&nbsp;alert(value); // 1, затем 2<br>
            }
        </code>
        <p>Выглядит гораздо красивее, чем использование <code>.next().value</code>, верно?</p>
        <p>... Но обратите внимание: пример выше выводит значение <code>1</code>, затем <code>2</code>. Значение
        <code>3</code> выведено не будет!</p>
        <p>Это из-за того, что перебор через <code>for...of</code> игнорирует последнее значение, при котором
        <code>done: true</code>. Поэтому, если мы хотим, чтобы были все значения при переборе <code>for...of</code>,
        то надо возвращать их через <code>yield</code>:</p>
        <code>
            function* generateSequence() {<br>
            &nbsp;&nbsp;yield 1;<br>
            &nbsp;&nbsp;yield 2;<br>
            &nbsp;&nbsp;yield 3;<br>
            }<br>
            <br>
            let generator = generateSequence();<br>
            <br>
            for (let value of generator) {<br>
            &nbsp;&nbsp;alert(value); // 1, затем 2, затем 3<br>
            }
        </code>
        <p>Так как генераторы являются перебираемыми объектами, мы можем использовать всю связанную с ними функциональность,
        например оператор расширения <code>...</code>:</p>
        <code>
            function* generateSequence() {<br>
            &nbsp;&nbsp;yield 1;<br>
            &nbsp;&nbsp;yield 2;<br>
            &nbsp;&nbsp;yield 3;<br>
            }<br>
            <br>
            let sequence = [0, ...generateSequence()]<br>
            <br>
            alert(sequence); // 0,1,2,3
        </code>
        <p>В коде выше <code>...generateSequence()</code> превращает перебираемый объект-генератор в массив элементов.</p>
    </article>

    <article>
        <h2>Использование генераторов для перебираемых объектов</h2>
        <p>Некоторое время назад, в главе «Перебираемые объекты», мы создали перебираемый объект <code>range</code>,
        который возвращает значения <code>from...to</code>.</p>
        <p>Давайте вспомним код:</p>
        <code>
            let range = {<br>
            &nbsp;&nbsp;from: 1,<br>
            &nbsp;&nbsp;to: 5<br>
            &nbsp;&nbsp;<br>
            &nbsp;&nbsp;// for...of range вызывает этот метод один раз в самом начале<br>
            &nbsp;&nbsp;[Symbol.iterator()] {<br>
            &nbsp;&nbsp;&nbsp;&nbsp;// ... Он возвращает перебираемый объект:<br>
            &nbsp;&nbsp;&nbsp;&nbsp;// далее for...of работает только с этим объектом, запрашивая следующие значения<br>
            &nbsp;&nbsp;&nbsp;&nbsp;return {<br>
            &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;current: this.from,<br>
            &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;last: this.to,<br>
            &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<br>
            &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;// next() вызывается при каждой итерации цикла for...of<br>
            &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;next() {<br>
            &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;// Нужно вернуть значение как объект {done: ..., value: ...}<br>
            &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;if (this.current <= this.last) {<br>
            &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;return {done: false, value: this.current++};<br>
            &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;} else {<br>
            &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;return {done: true}<br>
            &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;}<br>
            &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;}<br>
            &nbsp;&nbsp;&nbsp;&nbsp;};<br>
            &nbsp;&nbsp;}<br>
            };<br>
            <br>
            // При переборе объекта range будут выведены числа от range.from до range.to<br>
            alert([...range]); // 1,2,3,4,5
        </code>
        <p>Мы можем использовать функцию-генератор для итерации, указав ее в <code>Symbol.iterator</code>.</p>
        <p>Вот тот же <code>range</code>, но с гораздо более компактным итератором:</p>
        <code>
            let range = {<br>
            &nbsp;&nbsp;from: 1,<br>
            &nbsp;&nbsp;to: 5,<br>
            &nbsp;&nbsp;<br>
            &nbsp;&nbsp;*[Symbol.iterator]() { // Краткая запись для [Symbol.iterator]: function* ()<br>
            &nbsp;&nbsp;&nbsp;&nbsp;for (let value = this.from; value <= this.to; value++) {<br>
            &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;yield value;<br>
            &nbsp;&nbsp;&nbsp;&nbsp;}<br>
            &nbsp;&nbsp;}<br>
            };<br>
            <br>
            alert([...range]); // 1,2,3,4,5
        </code>
        <p>Это работает, потому что <code>range[Symbol.iterator]()</code> теперь возвращает генератор, и его методы -
        в точности то, что ожидает <code>for...of</code>:</p>
        <ul>
            <li>
                <p>У него есть метод <code>.next()</code>.</p>
            </li>
            <li>
                <p>Который возвращает значения в виде <code>{value: ..., done: true/false</code>.</p>
            </li>
        </ul>
        <p>Это не совпадение, конечно. Генераторы были добавлены в язык JavaScript, в частности, с целью упростить
        создание перебираемых объектов.</p>
        <p>Вариант с генератором намного короче, чем исходный вариант перебираемого <code>range</code>, и сохраняет
        те же функциональные возможности.</p>

        <article class="note">
            <h2>Генераторы могут генерировать бесконечно</h2>
            <p>В примерах выше мы генерировали конечные последовательности, но мы также можем сделать генератор,
            который будет возвращать значения бесконечно. Например, бесконечная последовательность псевдослучайных чисел.</p>
            <p>Конечно, нам потребуется <code>break</code> (или <code>return</code>) в цикле <code>for...of</code>
            по такому генератору, иначе цикл будет продолжаться бесконечно, и скрипт «зависнет».</p>
        </article>
    </article>

    <article>
        <h2>Композиция генераторов</h2>
        <p>Композиция генераторов - это особенная возможность генераторов, которая позволяет прозрачно «встраивать»
        генераторы друг в друга.</p>
        <p>Например, у нас есть функция для генерации последовательности чисел:</p>
        <code>
            function* generateSequence(start, end) {<br>
            &nbsp;&nbsp;for (let i = start, i <= end; i++) yield i;<br>
            }
        </code>
        <p>Мы хотели бы использовать ее при генерации более сложной последовательности:</p>
        <ul>
            <li>
                <p>Сначала цифры <code>0...9</code> (с кодами символов 48...57).</p>
            </li>
            <li>
                <p>За которыми следуют буквы в верхнем регистре <code>A...Z</code> (коды символов 65...90).</p>
            </li>
            <li>
                <p>За которыми следуют буквы алфавита <code>a...z</code> (коды символов 97...122).</p>
            </li>
        </ul>
        <p>Мы можем использовать такую последовательность для генерации паролей, выбирать символы из нее, но сначала
        ее нужно сгенерировать.</p>
        <p>В обычной функции, чтобы объединить результаты из нескольких других функций, мы вызываем их, сохраняем
        промежуточные результаты, а затем в конце их объединяем.</p>
        <p>Для генераторов есть особый синтаксис <code>yield*</code>, который позволяет «вкладывать» генераторы один
        в другой (осуществлять их композицию).</p>
        <p>Вот генератор с композицией:</p>
        <code>
            function* generateSequence(start, end) {<br>
            &nbsp;&nbsp;for (let i = start; i <= end; i++) yield i;<br>
            }<br>
            <br>
            function* generatePasswordCodes() {<br>
            &nbsp;&nbsp;// 0...9<br>
            &nbsp;&nbsp;yield* generateSequence(48, 57);<br>
            &nbsp;&nbsp;<br>
            &nbsp;&nbsp;// A...Z<br>
            &nbsp;&nbsp;yield* generateSequence(65, 90);<br>
            &nbsp;&nbsp;<br>
            &nbsp;&nbsp;// a...z<br>
            &nbsp;&nbsp;yield* generateSequence(97, 122);<br>
            }<br>
            <br>
            let str = '';<br>
            <br>
            for (let code of generatePasswordCodes()) {<br>
            &nbsp;&nbsp;str += String.fromCharCode(code);<br>
            }<br>
            <br>
            alert(str); // 0...9A...Za...z
        </code>
        <p>Директива <code>yield*</code> <i>делегирует</i> выполнение другому генератору. Этот термин означает, что
        <code>yield* gen</code> перебирает генератор <code>gen</code> и прозрачно направляет вывод наружу. Как если бы
        значения были сгенерированы внешним генератором.</p>
        <p>Результат - такой же, как если бы мы встроили код из вложенных генераторов:</p>
        <code>
            function* generateSequence(start, end) {<br>
            &nbsp;&nbsp;for (let i = start; i <= end; i++) yield i;<br>
            }<br>
            <br>
            function* generateAlphaNum() {<br>
            &nbsp;&nbsp;// yield* generateSequence(48, 57);<br>
            &nbsp;&nbsp;for (let i = 48; i <= 57; i++) yield i;<br>
            &nbsp;&nbsp;<br>
            &nbsp;&nbsp;// yield* generateSequence(65, 90);<br>
            &nbsp;&nbsp;for (let i = 65; i <= 90; i++) yield i;<br>
            &nbsp;&nbsp;<br>
            &nbsp;&nbsp;// yield* generateSequence(97, 122);<br>
            &nbsp;&nbsp;for (let i = 97; i <= 122; i++) yield i;<br>
            }<br>
            <br>
            let str = '';<br>
            <br>
            for (let code of generateAlphaNum) {<br>
            &nbsp;&nbsp;str += String.fromCharCode(code);<br>
            }<br>
            <br>
            alert(str); // 0...9A...Za...z
        </code>
        <p>Композиция генераторов - естественный способ вставлять вывод одного генератора в поток другого. Она не
        использует дополнительную память для хранения промежуточных результатов.</p>
    </article>

    <article>
        <h2>yield - дорога в обе стороны</h2>
        <p>До этого момента генераторы сильно напоминали перебираемые объекты, со специальным синтаксисом для генерации
        значений. Но на самом деле они намного мощнее и гибче.</p>
        <p>Все дело в том, что <code>yield</code> - дорога в обе стороны: он не только возвращает результат наружу,
        но и может передавать значение извне в генератор.</p>
        <p>Чтобы это сделать, нам нужно вызвать <code>generator.next(arg)</code> с аргументом. Этот аргумент становится
        результатом <code>yield</code>.</p>
        <p>Продемонстрируем это на примере:</p>
        <code>
            function* gen() {<br>
            &nbsp;&nbsp;// Передает вопрос во внешний код и ожидаем ответа<br>
            &nbsp;&nbsp;let result = yield '2 + 2 = ?'; // (*)<br>
            &nbsp;&nbsp;<br>
            &nbsp;&nbsp;alert(result);<br>
            }<br>
            <br>
            let generator = gen();<br>
            <br>
            let question = generator.next().value; // <-- yield возвращает значение<br>
            <br>
            generator.next(4); // --> передаем результат в генератор
        </code>
        <ol>
            <li>
                <p>Первый вызов <code>generator.next()</code> - всегда без аргумента, он начинает выполнение и возвращает
                результат первого <code>yield '2 + 2 = ?'</code>. На этой строчке генератор приостанавливает выполнение.</p>
            </li>
            <li>
                <p>Затем результат <code>yield</code> переходит во внешний код в переменную <code>question</code>.</p>
            </li>
            <li>
                <p>При <code>generator.next(4)</code> выполнение генератора возобновляется, а <code>4</code> выходит
                из присваивания как результат: <code>let result = 4</code>.</p>
            </li>
        </ol>
        <p>Обратите внимание, что внешний код не обязан немедленно вызывать <code>next(4)</code>. Ему может потребоваться
        время. Это не проблема, генератор подождет.</p>
        <p>Например:</p>
        <code>
            // Возобновить генератор через некоторое время<br>
            setTimeout(() => generator.next(4), 1000);
        </code>
        <p>Как видно, в отличие от обычных функций, генератор может обмениваться результатами с вызывающим кодом,
        передавая значения в <code>next/yield</code>.</p>
        <p>Чтобы сделать происходящее более очевидным, вот еще один пример с большим количеством вызовов:</p>
        <code>
            function* gen() {<br>
            &nbsp;&nbsp;let ask1 = yield '2 + 2 = ?';<br>
            &nbsp;&nbsp;alert(ask1); // 4<br>
            &nbsp;&nbsp;<br>
            &nbsp;&nbsp;let aks2 = yield '3 * 3 = ?';<br>
            &nbsp;&nbsp;alert(ask2); // 9<br>
            }<br>
            <br>
            let generator = gen();<br>
            <br>
            alert(generator.next().value); // '2 + 2 = ?'<br>
            alert(generator.next(4).value); // '3 * 3 = ?'<br>
            alert(generator.next(9).done); // true
        </code>
        <ol>
            <li>
                <p>Первый <code>.next()</code> начинает выполнение... Оно доходит до первого <code>yield</code>.</p>
            </li>
            <li>
                <p>Результат возвращается во внешний код.</p>
            </li>
            <li>
                <p>Второй <code>.next(4)</code> передает <code>4</code> обратно в генератор как результат первого
                <code>yield</code> и возобновляет выполнение.</p>
            </li>
            <li>
                <p>... Оно доходит до второго <code>yield</code>, который станет результатом <code>.next(4)</code>.</p>
            </li>
            <li>
                <p>Третий <code>next(9)</code> передает <code>9</code> в генератор как результат второго <code>yield</code>
                и возобновляет выполнение, которое завершается окончанием функции, так что <code>done: true</code>.</p>
            </li>
        </ol>
        <p>Получается такой «пинг-понг»: каждый <code>next(value)</code> передает в генератор значение, которое становится
        результатом текущего <code>yield</code>, возобновляет выполнение и получает выражение из следующего
        <code>yield</code>.</p>
    </article>

    <article>
        <h2>generator.throw</h2>
        <p>Как мы видели в примерах выше, внешний код может передавать значение в генератор как результат <code>yield</code>.</p>
        <p>... Но можно передать не только результат, но и инициировать ошибку. Это естественно, так как ошибка
        является своего рода результатом.</p>
        <p>Для того чтобы передать ошибку в <code>yield</code>, нам нужно вызвать <code>generator.throw(err)</code>.
        В таком случае исключение <code>err</code> возникнет на строке с <code>yield</code>.</p>
        <p>Например, здесь <code>yield '2 + 2 = ?'</code> приведет к ошибке:</p>
        <code>
            function* gen() {<br>
            &nbsp;&nbsp;try {<br>
            &nbsp;&nbsp;&nbsp;&nbsp;let result = yield '2 + 2 = ?'; // (1)<br>
            &nbsp;&nbsp;&nbsp;&nbsp;alert('Выполнение программы не дойдет до этой строки, потому что выше возникнет исключение');<br>
            &nbsp;&nbsp;} catch(err) {<br>
            &nbsp;&nbsp;&nbsp;&nbsp;alert(err); // Покажет ошибку<br>
            &nbsp;&nbsp;}<br>
            }<br>
            <br>
            let generator = gen();<br>
            <br>
            let question = generator.next().value;<br>
            <br>
            generator.throw(new Error('Ответ не найден в моей базе данных')); // (2)
        </code>
        <p>Ошибка, которая проброшена в генератор в строке <code>(2)</code>, приводит к исключению в строке
        <code>(1)</code> с <code>yield</code>. В примере выше <code>try...catch</code> перехватывает ее и отображает.</p>
        <p>Если мы не хотим перехватывать ее, то она, как и любое обычное исключение, «вывалится» из генератора
        во внешний код.</p>
        <p>Текущая строка вызывающего кода - это строка с <code>generator.throw</code>, отмечена <code>(2)</code>.
        Таким образом, мы можем отловить ее во внешнем коде, как здесь:</p>
        <code>
            function* generate() {<br>
            &nbsp;&nbsp;let result = yield '2 + 2 = ?'; // Ошибка в этой строке<br>
            }<br>
            <br>
            let generator = generate();<br>
            <br>
            let question = generator.next().value;<br>
            <br>
            try {<br>
            &nbsp;&nbsp;generator.throw(new Error('Ответ не найден в моей базе данных'));<br>
            } catch(err) {<br>
            &nbsp;&nbsp;alert(err); // Покажет ошибку<br>
            }
        </code>
        <p>Если же ошибка и там не перехвачена, то дальше - как обычно, она выпадет наружу и, если не перехвачена,
        «повалит» скрипт.</p>
    </article>

    <article>
        <h2>Итого</h2>
        <ul>
            <li>
                <p>Генераторы создаются при помощи функций-генераторов <code>function* f(...) {...}</code>.</p>
            </li>
            <li>
                <p>Внутри генераторов и только внутри них существует оператор <code>yield</code>.</p>
            </li>
            <li>
                <p>Внешний код и генератор обмениваются промежуточными результатами посредством вызовов
                <code>next/yield</code>.</p>
            </li>
        </ul>
        <p>В современном JavaScript генераторы используются редко. Но иногда они оказываются полезными, потому что
        способность функции обмениваться данными с вызывающим кодом во время выполнения совершенно уникальна.
        И, конечно, для создания перебираемых объектов.</p>
    </article>
</body>
</html>

<script src="script.js"></script>